cpp/src/parquet/column_reader.cc

```diff
@@ -306,8 +306,13 @@ class SerializedPageReader : public PageReader {
   // Please refer to the encryption specification for more details:
   // https://github.com/apache/parquet-format/blob/encryption/Encryption.md#44-additional-authenticated-data
 
-  // The ordinal fields in the context below are used for AAD suffix calculation.
+  // The CryptoContext used by this PageReader.
   CryptoContext crypto_ctx_;
+  // This PageReader has its own Decryptor instances in order to be thread-safe.
```

Comment:
Why not directly caching `meta_decryptor_` and `data_decryptor_` in the above `CryptoContext`?

Reply:
Yes, but isn't `CryptoContext` is created every time for a new `PageReader`? The old code has the problem because it shares decryptor instances from FileReader all the way down to PageReader. However, now we can create decryptors when creating the `CryptoContext` and cache them along with `CryptoContext`.

It is misleading that `PageReader::Open` takes `const CryptoContext*` but actually copies everything it passes.

Reply:
Yeah, everything is misleading in the Parquet encryption code. It's an entire mess.

Reply:
But, still, we should be careful that "copying" doesn't mean the actual objects are cloned when we've got `shared_ptr` members.

Besides, the nice thing with these factory functions is that they capture some precomputed encryption parameters such that calling them only creates a `Decryptor` instance (instead of fetching keys etc.).